nessie/src/main/java/org/apache/iceberg/nessie/NessieTableOperations.java

```diff
@@ -0,0 +1,151 @@
+  private static DynFields.UnboundField<Object> sparkConfMethod;
+  private static DynFields.UnboundField<Object> appIdMethod;
+  private final Configuration conf;
+  private final NessieClient client;
+  private final ContentsKey key;
+  private UpdateableReference reference;
+  private IcebergTable table;
+  private HadoopFileIO fileIO;
+
+  /**
+   * Create a nessie table operations given a table identifier.
+   */
+  public NessieTableOperations(
+      Configuration conf,
+      ContentsKey key,
+      UpdateableReference reference,
+      NessieClient client) {
+    this.conf = conf;
+    this.key = key;
+    this.reference = reference;
+    this.client = client;
+  }
+
+  @Override
+  protected void doRefresh() {
+    // break reference with parent (to avoid cross-over refresh)
+    // TODO, confirm this is correct behavior.
+    // reference = reference.copy();
+
+    reference.refresh();
+    String metadataLocation = null;
+    try {
+      Contents contents = client.getContentsApi().getContents(key, reference.getHash());
+      this.table = contents.unwrap(IcebergTable.class)
+          .orElseThrow(() ->
+              new IllegalStateException("Cannot refresh iceberg table: " +
+                  String.format("Nessie points to a non-Iceberg object for path: %s.", key)));
+      metadataLocation = table.getMetadataLocation();
+    } catch (NessieNotFoundException ex) {
+      if (currentMetadataLocation() != null) {
+        throw new NoSuchTableException(ex, "No such table %s", key);
+      }
+    }
+    refreshFromMetadataLocation(metadataLocation, 2);
+  }
+
+  @Override
+  protected void doCommit(TableMetadata base, TableMetadata metadata) {
+    reference.checkMutable();
+
+    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+
+    try {
+      IcebergTable newTable = ImmutableIcebergTable.builder().metadataLocation(newMetadataLocation).build();
+      client.getContentsApi().setContents(key,
+                                          reference.getAsBranch().getName(),
+                                          reference.getHash(),
+                                          String.format("iceberg commit%s", applicationId()),
+                                          newTable);
+    } catch (NessieConflictException ex) {
+      io().deleteFile(newMetadataLocation);
+      String fixMsg = reference.isBranch() ?
+          String.format("Update the reference %s and try again", reference.getName()) :
+          String.format("Can't commit to the tag %s", reference.getName());
+      throw new CommitFailedException(ex, "Commit failed: Reference hash is out of date. %s", fixMsg);
+    } catch (NessieNotFoundException ex) {
+      io().deleteFile(newMetadataLocation);
+      throw new RuntimeException(String.format("Commit failed: Reference %s does not exist", reference.getName()), ex);
+    } catch (Throwable e) {
+      io().deleteFile(newMetadataLocation);
```

Comment:
What about doing this delete in a `finally` block if `threw` is false? That's usually a better way than catching `Throwable` and wrapping it in a `RuntimeException`.

Reply:
agreed, fixed